### ISSUE TITLE
${java.home} does resolves to Eclipse JRE rather than target JRE for project

### ISSUE THREAD

**Issue by adriansuarez** [eclipse-m2e/m2e-core#248]
I have the following code snippet in my project:

```
        <dependency>
            <groupId>sun.jdk</groupId>
            <artifactId>tools</artifactId>
            <version>8</version>
            <scope>system</scope>
            <systemPath>${java.home}/../lib/tools.jar</systemPath>
        </dependency>
```

This generates an error because `${java.home}` resolves to the JRE used by Eclipse (Java 11) rather than the target JRE for the project (Java 8), and the Eclipse JRE does not have `tools.jar`. I was able to work around this by just copying `tools.jar` from the target JRE into the one that the plugin insists on using, but that shouldn't be necessary.

Apparently, this issue has been around since 2014: https://bugs.eclipse.org/bugs/show_bug.cgi?id=432992

**Comment by karypid** [comment]
Hello,

I am experiencing the issue described here. Is anyone working on this? How hard would it be to fix? I occassionaly contribute to Eclipse (especially for obscure bugs I am facing) but have a feeling that this may be something engrained into the way m2e works and not suitable for someone like me to pick up....

I have a feeling that the maven POM resolution executes with the embedded maven runtime inside the Eclipse JVM, hence Maven picks up that JVM as the target. Using an external Maven runtime simply loads the JARs of that runtime into the Eclipse JVM, so it does not solve the issue (it's a different Maven runtime, but still running within the Eclipse JVM).

So, would the only way to address this be to launch a maven runtime as an external process using the JRE configured for the project(s) involved? I suppose that's not easy as there must be a lot of integration code/hooks running alongside Maven while m2e invokes it which would require to be moved to IPC?

**Comment by laeubi** [comment]
The best way is to setup an IDE and start debugging the code, if there are specific questions let us know.

**Comment by karypid** [comment]
So, I was looking at this from a pure-maven perspective:

In official documentation for [Compiling Sources Using A Different JDK](https://maven.apache.org/plugins/maven-compiler-plugin/examples/compile-using-different-jdk.html#) they literally suggest to fork when running the compiler.

So not sure if anything can be done on m2e side to support this...

**Comment by karypid** [comment]
Some information after debugging to find the relevant code paths:

- The exit point from m2e code when resolving locations was modified recently and lies in [MavenExecutionContext](https://github.com/eclipse-m2e/m2e-core/commit/aa98a699e27ddaf5eed8a03e01817e8d07723a6c#diff-796a271293c65b5cac9c04a4dccddff344e43fd10f474a5cd2992c2333325704R294) where the call to `ensureDependenciesAreResolved` actuates the Maven part that tries to locate the dependency
- This ends up in Maven code which calls `getSystemPath()` in the maven model. The stack looks like this:
```
Thread [Worker-13: Building] (Suspended)	
	Dependency.getSystemPath() line: 553	
	MavenMetadataSource.createDependencyArtifact(ArtifactFactory, Dependency, String, ArtifactFilter) line: 364	
	MavenMetadataSource.createArtifacts(ArtifactFactory, List<Dependency>, String, ArtifactFilter, MavenProject) line: 518	
	DefaultProjectArtifactFactory.createArtifacts(MavenProject) line: 53	
	LifecycleDependencyResolver.resolveProjectDependencies(MavenProject, Collection<String>, Collection<String>, MavenSession, boolean, Set<Artifact>) line: 125	
	MojoExecutor.ensureDependenciesAreResolved(MojoDescriptor, MavenSession, DependencyContext) line: 401	
	MavenExecutionContext.executeMojo(MavenSession, MojoExecution) line: 317	
	MavenExecutionContext.lambda$0(MojoExecution, IMavenExecutionContext, IProgressMonitor) line: 278	
...
```

At this point the system path stored in that variable is already pointing to the location relative to the JDK that Eclipse is running with (for me version 17) whereas the project being compiled in the runtime workspace targets a JDK 8 installation in a different folder.

### A really hacky way to do this is to filter the returned values on m2e-side at this point

...and search for artifacts in system scope, that originally used `${java.home}` and use the eclipse project's JDK as the folder value, replacing the location of that maven model entry with the correct one.

Now, I set a modification breakpoint for that attribute and discovered that a simple String property interpolator replaces the ${java.home} with the actual directory location with the stack looking like this:

```
Thread [Worker-19: Updating Maven Project] (Suspended (breakpoint at line 712 in Dependency))	
	Dependency.setSystemPath(String) line: 712	
	StringVisitorModelInterpolator$ModelVisitor.visit(Dependency) line: 1281	
	StringVisitorModelInterpolator$ModelVisitor.visit(ModelBase) line: 1031	
	StringVisitorModelInterpolator$ModelVisitor.visit(Model) line: 170	
	StringVisitorModelInterpolator.interpolateModel(Model, File, ModelBuildingRequest, ModelProblemCollector) line: 107	
	DefaultModelBuilder.interpolateModel(Model, ModelBuildingRequest, ModelProblemCollector) line: 859	
	DefaultModelBuilder.build(ModelBuildingRequest, Collection<String>) line: 417	
	DefaultModelBuilder.build(ModelBuildingRequest) line: 267	
	DefaultProjectBuilder.build(List<ProjectBuildingResult>, List<InterimResult>, Map<String,MavenProject>, File, Set<File>, boolean, boolean, InternalConfig) line: 448	
	DefaultProjectBuilder.build(List<ProjectBuildingResult>, List<InterimResult>, Map<String,MavenProject>, List<File>, Set<File>, boolean, boolean, InternalConfig) line: 414	
	DefaultProjectBuilder.build(List<File>, boolean, ProjectBuildingRequest) line: 377	
	MavenImpl.readMavenProjects(Collection<File>, ProjectBuildingRequest) line: 480	
...
```

I haven't read the maven-side code in detail, but if there is any way to control the interpolator's variable-space from the m2e side, that would clearly be a better way to handle this. Maven would then inject the correct ${java.home} and be consistent throughout...

**Comment by karypid** [comment]
Apologies for the screenshot, but it seems that the interpolator uses a list of ValueSource objects. The one that resolves is at index 6 in the screenshot, which seems to contain all the calculated maven properties (env.ENV_VAR_NAME, system properties, etc).

I don't know if it's a system property that populates this value source, but overwriting java.home at this point to match the Eclipse project's target JRE would likely address this. It would be a bit inconsistent though to only address `java.home` in this way (e.g. why not `java.version` as well)...

![image](https://user-images.githubusercontent.com/1221101/177228335-75ee4571-9a2a-4b41-944d-07793b3bb069.png)

**Comment by karypid** [comment]
Ok, so INDEED we can control system properties value source from m2e side. As a proof-of-concept, I changed `org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.populateSystemProperties(MavenExecutionRequest)` to:

```
  public static void populateSystemProperties(MavenExecutionRequest request) {
    // temporary solution for https://issues.sonatype.org/browse/MNGECLIPSE-1607
    // oddly, there are no unit tests that fail if this is commented out
    if(request.getSystemProperties() == null || request.getSystemProperties().isEmpty()) {
      Properties systemProperties = new Properties();
      EnvironmentUtils.addEnvVars(systemProperties);
      copyProperties(systemProperties, System.getProperties());
      // ADDED THE LINE BELOW TO POINT TO MY LOCATION
      systemProperties.put("java.home", "C:\\Users\\karypid\\scoop\\apps\\temurin8-jdk\\current\\bin\\");
      MavenProperties.setProperties(systemProperties);
      request.setSystemProperties(systemProperties);
    }
  }
````

Now my project "works" in that (as you can see in my runtime workspace) my Eclipse is running with JDK 17 in `temurin-lts-jdk` directory, whereas the project behind my about dialog is targeting JDK 8 which is in directory `temurin8-dk` and is successfully resolved (no error in the pom and the maven dependencies points to the appropriate tools.jar).

![image](https://user-images.githubusercontent.com/1221101/177229434-9055eedd-8a12-4088-a00a-f0d28470493a.png)

I call this solution "LESS HACKY"

Would there be interest in a PR that adds/invokes a "populateTargetJDK()" method in `populateSystemProperties` so that I overwrite everything that is feasible with the values for the target JDK?

**Comment by karypid** [comment]
To clarify, it would look something like this::

```
  public static void populateSystemProperties(MavenExecutionRequest request) {
    // temporary solution for https://issues.sonatype.org/browse/MNGECLIPSE-1607
    // oddly, there are no unit tests that fail if this is commented out
    if(request.getSystemProperties() == null || request.getSystemProperties().isEmpty()) {
      Properties systemProperties = new Properties();
      EnvironmentUtils.addEnvVars(systemProperties);
      copyProperties(systemProperties, System.getProperties());
      populateTargetJDK(systemProperties, request.getProjectBuildingRequest().getProject());
      MavenProperties.setProperties(systemProperties);
      request.setSystemProperties(systemProperties);
    }
  }

  private static void populateTargetJDK(Properties systemProperties, MavenProject project) {
      if (chickenSwitch) return; // have some preference that controls this experimental feature...

      systemProperties.put("java.home", ...); // fetch from "project"
      systemProperties.put("java.version", ...); // fetch from "project"
      // ....
  }

```

**Comment by laeubi** [comment]
> In official documentation for [Compiling Sources Using A Different JDK](https://maven.apache.org/plugins/maven-compiler-plugin/examples/compile-using-different-jdk.html#) they literally suggest to fork when running the compiler.

m2e do not compile the code, JDT does ...

> To clarify, it would look something like this

we should not add any preference for this or will end up with a swtich for every request. The MavenExecutionFacade would be the place where we know the project and could get its JDK (possible needing some kind of extension point...)

**Comment by karypid** [comment]
I had a look at defining an extension point in the core plugin using the following interface:

```
public interface IMavenExecutionContextEnricher {
  default void populateSystemProperties(IMavenProjectFacade mpf, Properties systemProperties) {}
}
```

I was able to implement and contribute this in the m2e-jdt plugin so this seems like the way to go. But while working on this approach, I ran into another problem:

- In [deprecated code path](https://github.com/eclipse-m2e/m2e-core/blob/1.20.1/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/ProjectConfigurationManager.java#L325) there is a check that you only ever "configure" a single POM. This would have worked nicely with an extension that only has to worry about a specific project.
- The [active API for configuring that is used now](https://github.com/eclipse-m2e/m2e-core/blob/1.20.1/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/ProjectConfigurationManager.java#L361) works on multiple POM files at once. I assume this is because the user may be reconfiguring the entire workspace, or a list of selected projects in their workspace.

The problem now is that when dealing with multiple projects in the workspace, each could be using a different JDK (for example one uses 8, one uses 11 and one uses 17). This means they cannot share the same execution context as that can only specify a single JDK...

One approach would be to cluster the list of POMs "per-JDK" and loop over sets of of projects to configure (where each set uses the same JDK. This would make the "extension" look something like:

```
public interface IMavenExecutionContextEnricher {
  List<Set<IProject>> clusterForConfiguration(List<IProject> projectsToConfigure);
  void populateSystemProperties(IMavenProjectFacade mpf, Properties systemProperties);
}
```

So basically then the core would ask extensions to split the list of projects into sets that they can handle as groups. The JDT plugin would then split them into lists per JDK. The core then loops over the sets to trigger the configuration and the JDT plugin can augment each execution with a single JDK (using the first project in the active set of projects being configured).

Of course, one concern then becomes handling the case where multiple extensions may group in a different way. Perhaps we can make the extension point a 1-1 relationship between Core and JDT specifically, and then we don't need to worry about that?

**Comment by laeubi** [comment]
> IMavenExecutionContextEnricher

I think it might be more generic to have `IMavenExecutionRequestInitializer ` this could even be useful for other approaches!



> The problem now is that when dealing with multiple projects in the workspace, each could be using a different JDK (for example one uses 8, one uses 11 and one uses 17). This means they cannot share the same execution context as that can only specify a single JDK...

An execution context is usually not shared between different projects and is used to execute one project at a time.

**Comment by karypid** [comment]
Thanks. I will toy with the simple initializer listener approach after work and revert back with a PR (or more questions if there are issues).

**Comment by karypid** [comment]
I wired the JDT plugin to the core plugin (short-wired is probably more appropriate) and I am now able to overwrite system properties whenever an execution context is created.

I have however come up against a chicken-and-egg situation. To descire the proper value for `java.home` I need to know the JDK from the Java project. But this is set by the m2e-jdt plugin  in method [addJREClasspathContainer](https://github.com/eclipse-m2e/m2e-core/blob/1.20.1/org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/AbstractJavaProjectConfigurator.java#L218). 

The environment ID comes from the [configure](https://github.com/eclipse-m2e/m2e-core/blob/1.20.1/org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/AbstractJavaProjectConfigurator.java#L159) which looks at the options generated by [addJavaProjectOptions](https://github.com/eclipse-m2e/m2e-core/blob/595db9d68e6d090537a107e272d4a9c35a4213dc/org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/AbstractJavaProjectConfigurator.java#L583).

This uses method [getCompilerLevel](https://github.com/eclipse-m2e/m2e-core/blob/595db9d68e6d090537a107e272d4a9c35a4213dc/org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/AbstractJavaProjectConfigurator.java#L741) to find the Java version required by the compilation-related mojos which are calculated by [getCompilerMojoExecutions](https://github.com/eclipse-m2e/m2e-core/blob/595db9d68e6d090537a107e272d4a9c35a4213dc/org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/AbstractJavaProjectConfigurator.java#L735).

While prototypin, I converted most of these methods (`getCompilerLevel, getCompilerMojoExecutions, etc`) to static, as they do not really need the configurator state to work. I was therefore able to call them from my new `IMavenExecutionContextInitializer` class hoping to obtain the compiler level and then use that to set the properties appropriately.

This is where I ran into an endless recursion: to get the list of compiler-related mojos, m2e creates a new execution request, as it needs to invoke the maven runtime to get the execution plan (and then filter for the compiler-related stuff). But then this invokes my execution context initializer,. which in turn tries to detect the compiler level, which creates a new execution request, which calls my initializer again, and so on and so forth (stack overlow from infinite recursion)...

Basically I need to be able to look at maven-compiler-plugin to assess the source/target versions without creating a maven execution. Is that possible?

Alternatively, I noticed some hackery in [getExecutionPlans](https://github.com/eclipse-m2e/m2e-core/blob/595db9d68e6d090537a107e272d4a9c35a4213dc/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java#L408) which stores the execution plan in the maven project [at this point](https://github.com/eclipse-m2e/m2e-core/blob/595db9d68e6d090537a107e272d4a9c35a4213dc/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java#L413). I mention this because perhaps the core plugin should check whether the execution plan has been cached and only invoke the initializer to call `getCompilerLevel` if that is available. If not, the execution context would operate with the host-JDK (the one running Eclipse) until it gets to the point where an execution plan is avaialble...

I also thought of checking the java project to see if there is a classpath entry for the JRE in my initiallizer (idea was that I only populate the proper `java.home` when one is present) but does not work because the JRE classpath entry is added by the configuration change listener towards the end of the process. Therefore when the dependency resolution resolves the ${java.home} to point to a file, the JRE classpath entry.

Someone knowledgable with m2e needs to provide guidance on this as I'm not clear on how best to procced.

**Comment by laeubi** [comment]
I think it would be best if you could open a PR with your current code changes, that way one might give much better guidance to your problem.

**Comment by karypid** [comment]
So, I've pushed a branch in my fork to discuss. See: https://github.com/eclipse-m2e/m2e-core/pull/846/files

Basically the only reason this works is because I clear the initializer while it is running and restore it after it has run. Therefore, when the initializer itself creates a maven execution context for the purpose of resolving the maven-compiler-plugin version, that particular maven execution context is NOT augmented (the initializer does not invoke itself recursively) and is allowed to be used (to obtain the compiler version).

Clearly this is not thread-safe by the way, but as I mentioned in my previous comment: the hackery in [getExecutionPlans](https://github.com/eclipse-m2e/m2e-core/blob/595db9d68e6d090537a107e272d4a9c35a4213dc/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java#L408) which stores the execution plan in the maven project [at this point](https://github.com/eclipse-m2e/m2e-core/blob/595db9d68e6d090537a107e272d4a9c35a4213dc/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java#L413) is also [not thead-safe](https://github.com/eclipse-m2e/m2e-core/blob/595db9d68e6d090537a107e272d4a9c35a4213dc/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java#L420)...

**Comment by karypid** [comment]
@laeubi I've tried to use the declarative service approach (not familiar with these) so I've pushed a separate PR at https://github.com/eclipse-m2e/m2e-core/pull/847 

The problem is that I am doing something wrong and this causes m2e-jdt to fail to activate. What I did was:

1. Enable DS Annotations in project `org.eclipse.m2e.jdt`
2. Expose a service from MaventJdtPlugin
3. Reference the service from MavenImpl

Otherwise just rearranged the code a bit and moved it to the places you indicated.

I am not familiar with how declarative services initialize, but the net result is that while starting the JDT plugin is now no longer able to locate this service:

```
IMavenProjectRegistry projectManager = MavenPluginActivator.getDefault().getMavenProjectManager();
``` 

So this causes an NPE a few lines further down:

```
Caused by: java.lang.NullPointerException: Cannot invoke "org.eclipse.m2e.core.project.IMavenProjectRegistry.addMavenProjectChangedListener(org.eclipse.m2e.core.project.IMavenProjectChangedListener)" because "projectManager" is null
	at org.eclipse.m2e.jdt.MavenJdtPlugin.start(MavenJdtPlugin.java:144)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$2.run(BundleContextImpl.java:818)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$2.run(BundleContextImpl.java:1)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:810)
	... 208 more
```

I am pretty sure that I've probably caused the activation to occur earlier than previously by adding my service to m2e-jdt and at that point the `IMavenProjectRegistry` is not available yet. But clearly, I've not idea how DS work (never used them before) so if you can point me how to make this service provision become "advertised/registered/resolved late" (or whatever it is I need to do) would be able to make more progress.

Thanks!

**Comment by laeubi** [comment]
Don't use the static access with DS, you can simply do

```
@Reference
IMavenProjectRegistry projectManager;
```

and everything will be handeled auto-magically.

**Comment by karypid** [comment]
So, I did this but it still throws the NPE.

Looking around I found that the references seem to need XML definitions in OSGI-INF/<component>.xml. There are missing from the m2e-jdt plugin's  generated XML (even though I enabled DS Annotations with version 1.3). Any idea what I have  wrong?

What is missing is this sort of content:

```
   <reference cardinality="1..1" field="workspace" interface="org.eclipse.core.resources.IWorkspace" name="workspace"/>
   <reference cardinality="1..1" field="projectManager" interface="org.eclipse.m2e.core.project.IMavenProjectRegistry" name="projectManager"/>
   <reference cardinality="1..1" field="mavenConfiguration" interface="org.eclipse.m2e.core.embedder.IMavenConfiguration" name="mavenConfiguration"/>
```

**Comment by karypid** [comment]
Scratch that, adding `@Component` to MavenJdtPlugin fixed this but I still get the NPE. Will push an update now.

**Comment by karypid** [comment]
So the final fix for the declarative service was to stop implementing the Bundle start/stop methods and instead use `@Activate` and `@Deactivate` in the `MavenJdtPlugin` class.

I have pushed a PR that works and is more in line with @laeubi  instructions, but please see my own review comments as I still need help here.

**Comment by laeubi** [comment]
@karypid Good to see you making progress here, I'll try to review your code tomorrow in more detail.
